tutorial/20_recipes/013_wilcoxon_pruner.py

```diff
@@ -0,0 +1,324 @@
+# In practice, you should set a larger number of iterations.
+
+
+N_SA_ITER = 10000
+count = 0
+
+
+###################################################################################################
+# We counts the number of evaluation to know how many problems is pruned.
+
+
+num_evaluation = 0
+
+
+###################################################################################################
+# In this tutorial, we optimize three parameters: ``T0``, ``alpha``, and ``patience``.
+#
+# ``T0`` and ``alpha`` defining the temperature schedule
+# ---------------------------------------------------------------------------------------
+#
+# In simulated annealing, it is important to determine a good temperature scheduling, but
+# there is no "silver schedule" that is good for all problems, so we must tune the schedule
+# for this problem.
+# This code parametrizes the temperature as a monomial function ``T0 * (1 - t) ** alpha``, where
+# `t` progresses from 0 to 1. We try to optimize the two parameters ``T0`` and ``alpha``.
+#
+# ``patience``
+# -----------------------------
+#
+# This parameter specifies a threshold of how many iterations we allow the annealing process
+# continue without updating the best value. Practically, simulated annealing often drives
+# the solution far away from the current best solution, and rolling back to the best solution
+# periodically often improves optimization efficiency a lot. However, if the rollback happens
+# too often, the optimization may get stuck in a local optimum, so we must tune the threshold
+# to a sensible amount.
+#
+# .. note::
+#     Some samplers, including the default ``TPESampler``, currently cannot utilize the
+#     information of pruned trials effectively (especially when the last intermediate value
+#     is not the best approximation to the final objective function).
+#     As a workaround for this issue, you can return an estimation of the final value
+#     (e.g., the average of all evaluated values) when ``trial.should_prune()`` returns ``True``,
+#     instead of `raise optuna.TrialPruned()`.
+#     This will improve the sampler performance.
+
+
+def objective(trial: optuna.Trial) -> float:
+    global num_evaluation
+    options = SAOptions(
+        max_iter=N_SA_ITER,
+        T0=trial.suggest_float("T0", 0.01, 10.0, log=True),
+        alpha=trial.suggest_float("alpha", 1.0, 10.0, log=True),
+        patience=trial.suggest_int("patience", 10, 1000, log=True),
+    )
+    results = []
+
+    # For best results, shuffle the evaluation order in each trial.
+    ordering = np.random.permutation(len(dataset))
+    for i in ordering:
+        num_evaluation += 1
+        result_idxs = tsp_simulated_annealing(vertices=dataset[i], options=options)
+        result_cost = tsp_cost(dataset[i], result_idxs)
+        results.append(result_cost)
+
+        trial.report(result_cost, i)
+        if trial.should_prune():
+            # raise optuna.TrialPruned()
```

Comment:
```suggestion
            # This is a standard logic of pruning in Optuna, but this leads to degradation with TPE sampler.
            # raise optuna.TrialPruned()
```

Reply:
It not only degrades TPE Sampler, but also other samplers, so maybe we do not have to say TPESampler.